pkg/claims/migrateClaimsWrapper.go

```diff
@@ -35,33 +54,184 @@ func (w *migrateClaimsWrapper) Store() crud.Store {
 	return w.storeWithConnect
 }
 
-func (w *migrateClaimsWrapper) Read(itemType string, name string) ([]byte, error) {
-	// Read is also called during List (ReadAll) so this will migrate everything for a list too
-	return w.Migrate(itemType, name)
+func (w *migrateClaimsWrapper) List(itemType string, group string) ([]string, error) {
+	names, err := w.Store().List(itemType, group)
+	if err != nil {
+		if strings.Contains(err.Error(), crud.ErrRecordDoesNotExist.Error()) {
+
+			// Proactively migrate all data in the claims/ dir if we detect the layout is old
+			err := w.MigrateAll()
+			if err != nil {
+				return nil, err
+			}
+
+			return w.Store().List(itemType, group)
+
+		}
+
+		return nil, err
+	}
+
+	return names, nil
 }
 
-func (w *migrateClaimsWrapper) Migrate(itemType string, name string) ([]byte, error) {
+func (w *migrateClaimsWrapper) Read(itemType string, name string) ([]byte, error) {
 	data, err := w.Store().Read(itemType, name)
 	if err != nil {
+		// If we can't read the data, check for the data layout migration from rewriting the claims spec
+		if getSchemaVersion(data) != claim.CNABSpecVersion {
+
+			// Proactively migrate all data in the claims/ dir if we detect the layout is old
+			err := w.MigrateAll()
+			if err != nil {
+				return nil, err
+			}
+
+			return w.Store().Read(itemType, name)
+		}
+
 		return nil, err
 	}
 
-	//  We only migrate claim data, ignore the rest
-	if itemType != claim.ItemTypeClaims {
-		return data, nil
+	return data, nil
+}
+
+func (w *migrateClaimsWrapper) MigrateAll() error {
+	fmt.Fprint(w.Err, "!!! Migrating claims data to match the CNAB Claim spec https://cdn.cnab.io/schema/cnab-claim-1.0.0-DRAFT+b5ed2f3/claim.schema.json. This is a one-way migration !!!\n")
+
+	installationNames, err := w.Store().List(claim.ItemTypeClaims, "")
+	if err != nil {
+		return errors.Wrapf(err, "!!! Migration failed, unable to list installation names")
+	}
+
+	for _, installationName := range installationNames {
+		err = w.MigrateInstallation(installationName)
+		if err != nil {
+			fmt.Fprintf(w.Err, errors.Wrapf(err, "Error migrating installation %s. Skipping.\n", installationName).Error())
+		}
 	}
 
-	// If we can't migrate the data at any point, print an error message and
-	// return the original claim so that the entire operation isn't halted for
-	// remaining claims
+	fmt.Fprintf(w.Err, "!!! Migration complete !!!\n")
```

Comment:
Thoughts on only printing the migration details (here and elsewhere) if `porter-debug` is true?

Reply:
We used to print on  --debug but this is a critical migration that could go wrong and it's important for people to see it when it runs (it only runs once). If something goes wrong, you can't rerun with --debug on, you've missed it at that point.